### PR TITLE
Port upstream call tracer `withLog` to bsc

### PR DESCRIPTION
**Pull request by cannium**
### Description

Port https://github.com/ethereum/go-ethereum/pull/25991 to `bsc`, which adds `withLog` parameter to output event logs

### Rationale

We really need this event log output to fully trace all token transfers, both platform and BEP20 tokens. Please merge this patch or port it from upstream in your own way as soon as possible.

### Example & Changes

Please refer to the original patch https://github.com/ethereum/go-ethereum/pull/25991

### CODE REVIEW

**Review by hitdavid — APPROVED** [review]

**Comment by hitdavid** [comment]
This feature is already included in geth v1.11.0, and it is very useful.
Please merge it

**Comment by brilliant-lx** [comment]
Thx, it is ok to add this feature to BSC to keep the debug API consistent with Ethereum.

File: eth/tracers/internal/tracetest/calltrace_test.go [hunk 1]
```diff
@@ -382,7 +382,7 @@ func TestZeroValueToNotExitCall(t *testing.T) {
 	if err := json.Unmarshal(res, have); err != nil {
 		t.Fatalf("failed to unmarshal trace result: %v", err)
 	}
-	wantStr := `{"type":"CALL","from":"0x682a80a6f560eec50d54e63cbeda1c324c5f8d1b","to":"0x00000000000000000000000000000000deadbeef","value":"0x0","gas":"0x7148","gasUsed":"0x2d0","input":"0x","output":"0x","calls":[{"type":"CALL","from":"0x00000000000000000000000000000000deadbeef","to":"0x00000000000000000000000000000000000000ff","value":"0x0","gas":"0x6cbf","gasUsed":"0x0","input":"0x","output":"0x"}]}`
+	wantStr := `{"from":"0x682a80a6f560eec50d54e63cbeda1c324c5f8d1b","gas":"0x7148","gasUsed":"0x54d8","to":"0x00000000000000000000000000000000deadbeef","input":"0x","calls":[{"from":"0x00000000000000000000000000000000deadbeef","gas":"0x6cbf","gasUsed":"0x0","to":"0x00000000000000000000000000000000000000ff","input":"0x","value":"0x0","type":"CALL"}],"value":"0x0","type":"CALL"}`
```
Comment (brilliant-lx) [inline eth/tracers/internal/tracetest/calltrace_test.go:385]: why would`gasUsed` change? In this case: 0x2d0 -> 0x54d8?

File: eth/tracers/internal/tracetest/testdata/call_tracer/create.json [hunk 2]
```diff
@@ -48,7 +48,7 @@
   "result": {
     "from": "0x13e4acefe6a6700604929946e70e6443e4e73447",
     "gas": "0x5e106",
-    "gasUsed": "0x5e106",
+    "gasUsed": "0x897be",
```
Comment (brilliant-lx) [inline eth/tracers/internal/tracetest/testdata/call_tracer/create.json:51]: why gasUsed changed?

File: eth/tracers/internal/tracetest/calltrace_test.go (hunk 1 above)
Comment (cannium) [reply eth/tracers/internal/tracetest/calltrace_test.go:385]: I believe it's related to the upstream commit `0f4942214d76385c52c7917212331c9bad8b3566`, or https://github.com/ethereum/go-ethereum/pull/26048. I ported it as part of the patch.

File: eth/tracers/internal/tracetest/testdata/call_tracer/create.json (hunk 2 above)
Comment (cannium) [reply eth/tracers/internal/tracetest/testdata/call_tracer/create.json:51]: Same answer as above

**Review by brilliant-lx — DISMISSED** [review]

**Review by guagualvcha — DISMISSED** [review]

**Comment by brilliant-lx** [comment]
change the target branch from master to develop

**Review by brilliant-lx — APPROVED** [review]

**Review by GalaIO — APPROVED** [review]
LGTM

**Comment by aoliao** [comment]
我切换到develop后重新make geth，然后启动geth成功，使用web3js调用 debug_traceCall 方法，设置withLog:true，返回的数据中仍然没有EventLog，请问时哪里的问题？

geth启动参数：
geth --config ./config.toml --datadir ./node --txlookuplimit 0 --cache 86016 --rpc.allow-unprotected-txs --syncmode full --tries-verify-mode none --diffblock 5000 --rpc.txfeecap 0 --rpc.gascap 0 --http --http.corsdomain "*" --pruneancient=true --diffsync=true

web3js 代码：
web3.currentProvider.send({
        jsonrpc: '2.0',
        method:'debug_traceCall',
        params:[traceParams, "latest", {
            tracer: 'callTracer',
            tracerConfig: { withLog: true },
        }],
        id: new Date().getTime()
    }, (error, result) => {
        if(error) {
            console.log(error)
            return
        }
        fs.writeFileSync("trace.out", JSON.stringify(result))
        
    })

**Comment by SerK0** [comment]
Is callTracer works only with debug_traceTransaction?
I am trying to use it in debug_traceCall as:
  tracer: 'callTracer',
  tracerConfig: { withLog: true }
  
It doesn't work, no logs it output.

**Pull request merged**